divi/src/masternode.cpp

```diff
@@ -889,25 +919,19 @@ bool CMasternodeBroadcast::CheckInputs(int& nDoS) const
 
     LogPrint("masternode", "mnb - Accepted Masternode entry\n");
 
-    if (GetInputAge(vin) < MASTERNODE_MIN_CONFIRMATIONS) {
+    const auto* confBlock = GetCollateralConfirmationBlock();
+    if (confBlock == nullptr) {
         LogPrint("masternode","mnb - Input must have at least %d confirmations\n", MASTERNODE_MIN_CONFIRMATIONS);
         return false;
     }
+    collateralHeight = confBlock->nHeight - MASTERNODE_MIN_CONFIRMATIONS + 1;
```

Comment:
About the first comment, see my reply above.  About the second comment, that indeed makes sense.  I will update the code to cache the block hash instead, which is more robust against chain reorgs.

Reply:
I've updated the change now accordingly.  The `GetCollateralBlock` function now returns directly the block where the collateral was confirmed, not 14 blocks after - this is certainly less confusing.

And the cache is now directly on the hash of that block, so that we can verify if it is still contained in `chainActive` and update the cache if not (to handle reorgs).  Since the expensive part is looking up the block from transaction hash, the cache will still help with that.